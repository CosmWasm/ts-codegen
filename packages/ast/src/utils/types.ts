import * as t from '@babel/types';
import { camel, pascal } from 'case';
import { propertySignature } from './babel';
import { RenderContext } from '../context';
import { JSONSchema } from '@cosmology/ts-codegen-types';

export function getResponseType(
  context: RenderContext,
  underscoreName: string
) {
  const methodName = camel(underscoreName);
  return pascal(
    context.contract?.responses?.[underscoreName]?.title
    ??
    // after v1.1 is adopted, we can deprecate this and require the above response
    `${methodName}Response`
  );
};

const getTypeStrFromRef = ($ref: string): string => {
  if ($ref?.startsWith('#/definitions/')) {
    return $ref.replace('#/definitions/', '');
  }
  throw new Error('what is $ref: ' + $ref);
}

export const getTypeFromRef = ($ref: string) => {
  return t.tsTypeReference(t.identifier(getTypeStrFromRef($ref)))
}

const getArrayTypeFromRef = ($ref: string) => {
  return t.tsArrayType(
    getTypeFromRef($ref)
  );
}


export const getTypeOrRef = (obj: JSONSchema): t.TSType => {
  if (obj.$ref) {
    return getTypeFromRef(obj.$ref);
  } else if (typeof obj.type === 'string') {
    return getType(obj.type);
  } else if (Array.isArray(obj.type)) {
    // Handle array of types by creating a union type
    console.log(obj.type);
    const types = obj.type.map(t => getType(t));
    return t.tsUnionType(types);
  } else {
    throw new Error('Type definition not supported or missing: ' + JSON.stringify(obj));
  }
};

const getArrayTypeFromItems = (
  context: RenderContext,
  items: JSONSchema | JSONSchema[]
): t.TSArrayType => {
  if (Array.isArray(items)) {
    // if (context.options.types.itemsUseTuples === true) {
    // Using the modified method with tsTupleType
    // return t.tsArrayType(
    // t.tsTupleType(items.map(item => getTypeOrRef(item)))
    // );
    // } else {
    // Using the original method with nested arrays
    if (items.length > 0 && items[0].type === 'array') {
      return getArrayTypeFromItems(context, items[0]);
    }
    return items.length > 0 ? t.tsArrayType(t.tsArrayType(getTypeOrRef(items[0]))) : t.tsArrayType(t.tsArrayType(t.tsAnyKeyword()));
    // }
  } else {
    const detect = detectType(items.type);

    if (detect.type === 'array') {
      if (Array.isArray(items.items)) {
        return t.tsArrayType(
          t.tsArrayType(
            getTypeOrRef(items.items[0])
          )
        );
      } else if (items.items) {
        return t.tsArrayType(
          getArrayTypeFromItems(context, items.items)
        );
      } else {
        // Handle the case where items.items is undefined for array types
        return t.tsArrayType(t.tsAnyKeyword())
      }
    }

    return t.tsArrayType(
      getType(detect.type)
    );
  }
};

export const detectType = (type: string | string[]) => {
  let optional = false;
  let theType = '';
  if (Array.isArray(type)) {
    if (type.length !== 2) {
      throw new Error('[getType(array length)] case not handled by transpiler. contact maintainers.')
    }
    const [nullableType, nullType] = type;
    if (nullType !== 'null') {
      throw new Error('[getType(null)] case not handled by transpiler. contact maintainers.')
    }
    theType = nullableType;
    optional = true;
  } else {
    theType = type;
  }

  return {
    type: theType,
    optional
  };
}

export const getTypeInfo = (
  context: RenderContext,
  info: JSONSchema
) => {
  let type = undefined;
  let optional = undefined;

  // Handle nullable types using 'anyOf' schema
  if (Array.isArray(info.anyOf)) {
    if (info.anyOf.length !== 2) {
      throw new Error('case not handled by transpiler. contact maintainers.');
    }
    const [nullableType, nullType] = info.anyOf;
    if (nullType?.type !== 'null') {
      throw new Error('[nullableType.type]: case not handled by transpiler. contact maintainers.');
    }
    if (!nullableType?.$ref) {
      if (nullableType.title) {
        type = t.tsTypeReference(t.identifier(nullableType.title));
      } else {
        throw new Error('[nullableType.title] case not handled by transpiler. contact maintainers.');
      }
    } else {
      type = getTypeFromRef(nullableType.$ref);
    }
    optional = true;
  } else if (typeof info.type === 'string') {
    console.log('is string')
    // Process type when it's explicitly defined
    if (info.type === 'array') {
      if (typeof info.items === 'object' && !Array.isArray(info.items)) {
        if (info.items.$ref) {
          type = getArrayTypeFromRef(info.items.$ref);
        } else if (info.items.title) {
          type = t.tsArrayType(t.tsTypeReference(t.identifier(info.items.title)));
        } else if (info.items.type) {
          type = getArrayTypeFromItems(context, info.items);
        } else {
          throw new Error('[info.items] case not handled by transpiler. contact maintainers.');
        }
        // type = getArrayTypeFromItems(context, info.items);
      } else {
        if (Array.isArray(info.items)) {
          type = getArrayTypeFromItems(context, info.items);
        } else {
          throw new Error('[info.items] case not handled by transpiler. contact maintainers.');
        }
      }
    } else {
      const detect = detectType(info.type);
      type = getType(detect.type);
      optional = detect.optional;
    }
  } else if (Array.isArray(info.type)) {
    console.log('is arr', info.type)
    // Handle multiple types, typically nullable types
    if (info.type.length !== 2) {
      throw new Error('please report this to maintainers (field type): ' + JSON.stringify(info, null, 2));
    }
    const [nullableType, nullType] = info.type;
    if (nullType !== 'null') {
      throw new Error('please report this to maintainers (field type): ' + JSON.stringify(info, null, 2));
    }

    if (nullableType === 'array' && typeof info.items === 'object' && !Array.isArray(info.items)) {
      if (info.items.type) {
        const detect = detectType(info.items.type);
        if (detect.type === 'array') {
          console.log('items type')
          type = t.tsArrayType(getArrayTypeFromItems(context, info.items));
        } else {
          type = t.tsArrayType(getType(detect.type));
        }
        optional = detect.optional;
      } else if (info.items.$ref) {
        type = getArrayTypeFromRef(info.items.$ref);
      } else if (info.items.type) {
        type = getArrayTypeFromItems(context, info.items);
      } else {
        throw new Error('[info.items] case not handled by transpiler. contact maintainers.');
      }
    } else {
      const detect = detectType(nullableType);
      optional = detect.optional;
      if (detect.type === 'array') {
        type = getArrayTypeFromItems(context, info.items);
      } else {
        type = getType(detect.type);
      }
    }

    optional = true;  // Ensure optional is set for nullable types
  }
  console.log({ type })

  return {
    type,
    optional
  };
};

export const getType = (type: string) => {
  switch (type) {
    case 'string':
      return t.tsStringKeyword();
    case 'boolean':
      return t.tSBooleanKeyword();
    case 'integer':
      return t.tsNumberKeyword();
    case 'null':
      return t.tsNullKeyword();
    case 'number':
      return t.tsNumberKeyword();
    default:
      throw new Error('contact maintainers [unknown type]: ' + type);
  }
}

export const getPropertyType = (
  context: RenderContext,
  schema: JSONSchema,
  prop: string
) => {
  const props = schema.properties ?? {};
  let info = props[prop];

  let type = null;
  let optional = !schema.required?.includes(prop);

  if (info.allOf && info.allOf.length === 1) {
    info = info.allOf[0];
  }

  if (typeof info.$ref === 'string') {
    type = getTypeFromRef(info.$ref)
  }

  const typeInfo = getTypeInfo(context, info);
  if (typeof typeInfo.optional !== 'undefined') {
    optional = typeInfo.optional;
  }
  if (typeof typeInfo.type !== 'undefined') {
    console.log(JSON.stringify(typeInfo, null, 2))
    type = typeInfo.type;
  }

  if (!type) {
    throw new Error('cannot find type for ' + JSON.stringify(info))
  }

  if (schema.required?.includes(prop)) {
    optional = false;
  }

  return { type, optional };
};


export function getPropertySignatureFromProp(
  context: RenderContext,
  jsonschema: JSONSchema,
  prop: string,
  camelize: boolean
) {
  if (jsonschema.properties[prop].type === 'object') {
    if (jsonschema.properties[prop].title) {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsTypeReference(t.identifier(jsonschema.properties[prop].title))
        )
      );
    } else {
      throw new Error('getPropertySignatureFromProp() contact maintainer');
    }
  }

  if (Array.isArray(jsonschema.properties[prop].allOf)) {
    const isOptional = !jsonschema.required?.includes(prop);
    const unionTypes = jsonschema.properties[prop].allOf.map(el => {
      if (el.title) return el.title;
      if (el.$ref) return getTypeStrFromRef(el.$ref);
      return el.type;
    }).filter(a => typeof a === 'string');
    // ONLY SUPPORT string types for now!
    // e.g. NOT supporting nullable types or type: string[] yet

    // @ts-ignore:next-line
    const uniqUnionTypes = [...new Set(unionTypes)];

    if (uniqUnionTypes.length === 1) {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsTypeReference(
            t.identifier(uniqUnionTypes[0] as string)
          )
        ),
        isOptional
      );
    } else {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsUnionType(
            uniqUnionTypes.map(typ =>
              t.tsTypeReference(
                t.identifier(typ as string)
              )
            )
          )
        ),
        isOptional
      );
    }
  } else if (Array.isArray(jsonschema.properties[prop].oneOf)) {
    const isOptional = !jsonschema.required?.includes(prop);
    const unionTypes = jsonschema.properties[prop].oneOf.map(el => {
      if (el.title) return el.title;
      if (el.$ref) return getTypeStrFromRef(el.$ref);
      return el.type;
    }).filter(a => typeof a === 'string');
    // ONLY SUPPORT string types for now!
    // e.g. NOT supporting nullable types or type: string[] yet

    const uniqUnionTypes = [...new Set(unionTypes)];
    if (uniqUnionTypes.length === 1) {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsTypeReference(
            t.identifier(uniqUnionTypes[0] as string)
          )
        ),
        isOptional
      );
    } else {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsUnionType(
            uniqUnionTypes.map(typ =>
              t.tsTypeReference(
                t.identifier(typ as string)
              )
            )
          )
        ),
        isOptional
      );
    }

  }

  try {
    getPropertyType(context, jsonschema, prop);
  } catch (e) {
    console.log(e);
    console.log(JSON.stringify(jsonschema, null, 2), prop);
  }

  const { type, optional } = getPropertyType(context, jsonschema, prop);
  return propertySignature(
    camelize ? camel(prop) : prop,
    t.tsTypeAnnotation(
      type
    ),
    optional
  );
}

export const getParamsTypeAnnotation = (
  context: RenderContext,
  jsonschema: any,
  camelize: boolean = true
): t.TSTypeAnnotation => {
  const keys = Object.keys(jsonschema.properties ?? {});

  if (!keys.length && jsonschema.$ref) {
    return t.tsTypeAnnotation(getTypeFromRef(jsonschema.$ref))
  }

  if (!keys.length) return undefined;

  const typedParams = keys.map(prop => getPropertySignatureFromProp(
    context,
    jsonschema,
    prop,
    camelize
  ));

  return t.tsTypeAnnotation(
    t.tsTypeLiteral(
      // @ts-ignore:next-line
      [
        ...typedParams
      ]
    )
  )
}

export const createTypedObjectParams = (
  context: RenderContext,
  jsonschema: JSONSchema,
  camelize: boolean = true
): (t.Identifier | t.Pattern | t.RestElement) => {

  const keys = Object.keys(jsonschema.properties ?? {});
  if (!keys.length) {
    // is there a ref?
    if (jsonschema.$ref) {
      const obj = context.refLookup(jsonschema.$ref);
      // If there is a oneOf, then we need to create a type for it
      if (obj?.oneOf) {
        // the actual type of the ref
        const refType = jsonschema.$ref.split('/').pop();
        const refName = camel(refType);
        const id = t.identifier(refName);
        id.typeAnnotation = t.tsTypeAnnotation(t.tsTypeReference(t.identifier(refType)));
        return id
      } else if (obj) {
        return createTypedObjectParams(
          context,
          obj,
          camelize
        );
      }
    }

    // no results...
    return;
  }

  const params = keys.map(prop => {
    return t.objectProperty(
      camelize ? t.identifier(camel(prop)) : t.identifier(prop),
      camelize ? t.identifier(camel(prop)) : t.identifier(prop),
      false,
      true
    );
  });

  const obj = t.objectPattern(
    [
      ...params
    ]
  );

  obj.typeAnnotation = getParamsTypeAnnotation(
    context,
    jsonschema,
    camelize
  );

  return obj;
};