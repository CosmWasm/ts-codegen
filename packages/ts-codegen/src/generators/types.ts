import { pascal } from "case";
import { header } from '../utils/header';
import { join } from "path";
import { sync as mkdirp } from "mkdirp";
import * as t from '@babel/types';
import { writeFileSync } from 'fs';
import generate from "@babel/generator";
import { clean } from "../utils/clean";
import { findAndParseTypes, findExecuteMsg, findQueryMsg } from '../utils';
import { ContractInfo, RenderContext, TSTypesOptions } from "wasm-ast-types";
import { BuilderFile } from "../builder";

export default async (
  name: string,
  contractInfo: ContractInfo,
  outPath: string,
  tsTypesOptions?: TSTypesOptions
): Promise<BuilderFile[]> => {

  const { schemas } = contractInfo;
  const context = new RenderContext(contractInfo, {
    types: tsTypesOptions ?? {}
  });
  const options = context.options.types;

  const localname = pascal(name) + '.types.ts';
  const ExecuteMsg = findExecuteMsg(schemas);
  const typeHash = await findAndParseTypes(schemas);

  const body = [];

  // TYPES
  Object.values(typeHash).forEach((type: t.Node) => {
    body.push(
      clean(type)
    )
  });

  // alias the ExecuteMsg (deprecated option)
  if (options.aliasExecuteMsg && ExecuteMsg) {
    body.push(
      t.exportNamedDeclaration(
        t.tsTypeAliasDeclaration(
          t.identifier(`${name}ExecuteMsg`),
          null,
          t.tsTypeReference(t.identifier('ExecuteMsg'))
        )
      )
    );
  }

  function addEntryPointAlias(msgName: string) {
    body.push(
      t.exportNamedDeclaration(
        t.tsTypeAliasDeclaration(
          t.identifier(`${name}${msgName}`),
          null,
          t.tsTypeReference(t.identifier(msgName))
        )
      )
    );
  }

  if (options.aliasEntryPoints) {
    if (ExecuteMsg) {
      addEntryPointAlias('ExecuteMsg');
    }
    if (findQueryMsg(schemas)) {
      addEntryPointAlias('QueryMsg');
    }
  }


  const imports = context.getImports();
  const code = header + generate(
    // @ts-ignore
    t.program([
      ...imports,
      ...body
    ])
  ).code;

  mkdirp(outPath);
  const filename = join(outPath, localname);
  writeFileSync(filename, code);

  return [
    {
      type: 'type',
      contract: name,
      localname,
      filename,
    }
  ]
};
