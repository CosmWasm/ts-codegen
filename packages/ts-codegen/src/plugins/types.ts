import * as t from '@babel/types';
import {
  ContractInfo,
  RenderContext,
  RenderContextBase,
  RenderOptions
} from '@cosmwasm/ts-codegen-ast';
import { pascal } from 'case';

import { BuilderFileType } from '../builder';
import { findAndParseTypes, findExecuteMsg } from '../utils';
import { clean } from '../utils/clean';
import { BuilderPluginBase } from './plugin-base';

export class TypesPlugin extends BuilderPluginBase<RenderOptions> {
  initContext(
    contract: ContractInfo,
    options?: RenderOptions
  ): RenderContextBase<RenderOptions> {
    return new RenderContext(contract, options, this.builder?.builderContext);
  }

  async doRender(
    name: string,
    context: RenderContext
  ): Promise<
    {
      type: BuilderFileType;
      pluginType?: string;
      localname: string;
      body: any[];
    }[]
  > {
    const { enabled } = this.options.types;

    if (!enabled) {
      return;
    }

    const { schemas } = context.contract;
    const options = this.options.types;

    const localname = pascal(name) + '.types.ts';
    const ExecuteMsg = findExecuteMsg(schemas);
    const typeHash = await findAndParseTypes(schemas);

    const body = [];

    // TYPES
    Object.values(typeHash).forEach((type: t.Node) => {
      body.push(clean(type));
    });

    // alias the ExecuteMsg
    if (options.aliasExecuteMsg && ExecuteMsg) {
      body.push(
        t.exportNamedDeclaration(
          t.tsTypeAliasDeclaration(
            t.identifier(`${name}ExecuteMsg`),
            null,
            t.tsTypeReference(t.identifier('ExecuteMsg'))
          )
        )
      );
    }

    return [
      {
        type: 'type',
        localname,
        body
      }
    ];
  }
}
