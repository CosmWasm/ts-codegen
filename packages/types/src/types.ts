export declare type fieldType = 'Long' | 'Coin' | 'Duration' | 'Height' | string;
export interface Field {
  name: string;
  type: fieldType;
  node: any;
};
export interface Interface {
  name: string;
  fields: Field[]
};

export interface QueryMsg {
  $schema: string;
  title: 'QueryMsg';
  oneOf?: any;
  allOf?: any;
  anyOf?: any;
}

export interface ExecuteMsg {
  $schema: string;
  title: 'ExecuteMsg' | 'ExecuteMsg_for_Empty';
  oneOf?: any;
  allOf?: any;
  anyOf?: any;
}

export interface JSONSchema {
  $ref?: string;
  $schema?: string;
  additionalProperties?: boolean;
  allOf?: JSONSchema[];
  anyOf?: JSONSchema[];
  definitions?: Record<string, JSONSchema>;
  description?: string;
  oneOf?: JSONSchema[];
  properties?: Record<string, JSONSchema>;
  patternProperties?: Record<string, JSONSchema>;
  items?: JSONSchema | JSONSchema[];
  additionalItems?: JSONSchema;
  required?: string[];
  title?: string;
  type?: string | string[]
}


// this is actually what items ends up being most of the time:
export type JSONSchemaType =
  | { type: 'integer'; format?: 'int32' | 'int64' }
  | { type: 'number'; format?: 'float' | 'double' }
  | { type: 'string'; format?: 'date-time' | 'email' | 'hostname' | 'ipv4' | 'ipv6' | 'uri' }
  | { type: 'boolean' }
  | { type: 'object'; properties?: { [key: string]: JSONSchemaType }; required?: string[] }
  | { type: 'array'; items?: JSONSchemaType | JSONSchemaType[]; maxItems?: number; minItems?: number }
  | JSONSchemaType[]; // For tuple definitions
