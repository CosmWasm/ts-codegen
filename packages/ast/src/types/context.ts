import { IDLObject, JSONSchema } from "@cosmwasm/ts-codegen-types";
import { MessageBuilderOptions, MessageComposerOptions, ReactQueryOptions, RecoilOptions, TSClientOptions, TSTypesOptions, UseContractsOptions } from "./plugins";

export interface ContractInfo {
  schemas: JSONSchema[];
  responses?: Record<string, JSONSchema>;
  idlObject?: Partial<IDLObject>;
};
export interface RenderOptions {
  enabled?: boolean;
  types?: TSTypesOptions;
  recoil?: RecoilOptions;
  messageComposer?: MessageComposerOptions;
  messageBuilder?: MessageBuilderOptions;
  client?: TSClientOptions;
  reactQuery?: ReactQueryOptions;
  useContractsHook?: UseContractsOptions;
}

export interface ProviderInfo {
  classname: string,
  filename: string,
  basename: string,
}
