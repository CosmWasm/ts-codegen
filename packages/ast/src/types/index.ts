export * from './context';
export * from './plugins';