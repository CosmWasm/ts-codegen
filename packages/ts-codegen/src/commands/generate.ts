import { prompt } from '../utils/prompt';
import codegen from '../index';
import { TSBuilderOptions } from '../builder';

export default async (argv) => {
    const questions = [
        {
            _: true,
            type: 'path',
            name: 'schema',
            message: 'which directory contains the the Rust contracts?',
            default: './schema'
        },
        {
            _: true,
            type: 'path',
            name: 'out',
            message: 'where is the output directory?',
            default: './ts'
        },
        {
            _: true,
            type: 'string',
            name: 'name',
            message: 'contract name?'
        },
        {
            type: 'checkbox',
            name: 'plugin',
            message: 'which plugins?',
            choices: [
                'client',
                'recoil',
                'react-query',
                'message-composer',
                'msg-builder',
                'abstract-app'
            ]
        },
        {
            type: 'confirm',
            name: 'bundle',
            message: 'enable bundle?',
            default: true
        }
    ];

    if (argv.typesOnly) {
        argv.plugin = 'types';
    }

    let { schema, out, name, plugin, bundle } = await prompt(questions, argv);
    if (!Array.isArray(plugin)) plugin = [plugin];

    ///////// REACT QUERY
    const questions2 = [];
    if (plugin.includes('react-query')) {
        [].push.apply(questions2, [
            {
                type: 'confirm',
                name: 'optionalClient',
                message: 'optionalClient?',
                default: false
            },
            {
                type: 'list',
                name: 'version',
                message: 'which react-query version?',
                default: 'v3',
                choices: ['v3', 'v4']
            },
            {
                type: 'confirm',
                name: 'queryKeys',
                message: 'queryKeys?',
                default: false
            },
        ])
    };
    const { optionalClient, version, queryKeys } = await prompt(questions2, argv);
    const questions3 = [];
    if (version === 'v4') {
        [].push.apply(questions3, [
            // currently we only support v4 for useMutation
            {
                type: 'confirm',
                name: 'mutations',
                message: 'Generate useMutation hooks?',
                default: false
            }

        ])
    };
    const { mutations } = await prompt(questions3, argv);

    const queryFactoryQuestions  = [];
    if (queryKeys) {
      [].push.apply(queryFactoryQuestions, [
        // Only can use queryFactory if queryKeys is enabled
        {
          type: 'confirm',
          name: 'queryFactory',
          message: 'queryFactory?',
          default: false
        }
      ])
    };
    const { queryFactory } = await prompt(queryFactoryQuestions, argv);
    ///////// END REACT QUERY

    ///////// BUNDLE
    const questions4 = [];
    if (bundle) {
        [].push.apply(questions4, [
            {
                type: 'string',
                name: 'bundleFile',
                message: 'bundleFile?',
                default: 'index.ts'
            },
            {
                type: 'string',
                name: 'bundleScope',
                message: 'bundleScope?',
                default: 'contracts'
            }
        ])
    };
    const { bundleFile, bundleScope } = await prompt(questions4, argv);
    ///////// END BUNDLE

    const options: TSBuilderOptions = {
        types: {
            enabled: true
        },
        client: {
            enabled:
                plugin.includes('client') ||
                plugin.includes('recoil') ||
              // react-query either uses client or abstract-app
              (plugin.includes('react-query') && !plugin.includes('abstract-app'))
        },
        reactQuery: {
            enabled: plugin.includes('react-query'),
            optionalClient,
            queryKeys,
            version,
            mutations,
            queryFactory
        },
        recoil: {
            enabled: plugin.includes('recoil'),
        },
        messageComposer: {
            enabled: plugin.includes('message-composer')
        },
        msgBuilder: {
          enabled: plugin.includes('msg-builder') || plugin.includes('abstract-app')
        },
        bundle: {
            enabled: bundle,
            scope: bundleScope,
            bundleFile
        },
        abstractApp: {
          enabled: plugin.includes('abstract-app'),
        }
    };

    await codegen({
        contracts: [
            {
                name,
                dir: schema
            }
        ],
        outPath: out,
        options
    })
};
