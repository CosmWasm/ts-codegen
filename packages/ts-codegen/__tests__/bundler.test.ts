import generate from '@babel/generator';
import * as t from '@babel/types';
import cases from 'jest-in-case'

import {
  BundleData,
  recursiveModuleBundle
} from '../src/bundler';

const preview = (ast: any) => {
  // @ts-ignore
  return generate(t.program(ast)).code;
}

interface JestBundleData {
    name: string;
    data: BundleData
}

cases(`recursiveModuleBundle`, (opts: JestBundleData) => {
  expect(preview(recursiveModuleBundle(opts.data))).toMatchSnapshot();
}, [
  {
    name: 'root',
    data: {
      osmosis: {
        __export: true,
        _0: true
      },
      tendermint: {
        __export: true,
        _0: true
      }
    }
  },
  {
    name: 'single',
    data: {
      ics23: {
        __export: true,
        _0: true
      }
    }
  },
  {
    name: 'tendermint',
    data: {
      tendermint: {
        abci: {
          __export: true,
          _223: true
        },
        crypto: {
          __export: true,
          _224: true,
          _225: true
        },
        libs: {
          bits: {
            __export: true,
            _226: true
          }
        },
        p2p: {
          __export: true,
          _227: true
        },
        types: {
          __export: true,
          _228: true,
          _229: true,
          _230: true,
          _231: true,
          _232: true
        },
        version: {
          __export: true,
          _233: true
        }
      }
    }
  },
  {
    name: 'osmo',
    data: {
      osmosis: {
        claim: {
          v1beta1: {
            __export: true,
            _177: true,
            _178: true,
            _179: true,
            _180: true
          }
        },
        epochs: {
          v1beta1: {
            __export: true,
            _181: true,
            _182: true
          }
        },
        gamm: {
          v1beta1: {
            __export: true,
            _183: true,
            _184: true,
            _185: true,
            _186: true,
            _187: true
          }
        },
        incentives: {
          __export: true,
          _188: true,
          _189: true,
          _190: true,
          _191: true,
          _192: true
        },
        lockup: {
          __export: true,
          _193: true,
          _194: true,
          _195: true,
          _196: true
        },
        mint: {
          v1beta1: {
            __export: true,
            _197: true,
            _198: true,
            _199: true
          }
        },
        poolincentives: {
          v1beta1: {
            __export: true,
            _200: true,
            _201: true,
            _202: true,
            _203: true
          }
        },
        store: {
          v1beta1: {
            __export: true,
            _204: true
          }
        },
        superfluid: {
          __export: true,
          _205: true,
          _206: true,
          _207: true,
          _208: true,
          _209: true
        },
        txfees: {
          v1beta1: {
            __export: true,
            _210: true,
            _211: true,
            _212: true,
            _213: true
          }
        }
      }
    }
  }
])