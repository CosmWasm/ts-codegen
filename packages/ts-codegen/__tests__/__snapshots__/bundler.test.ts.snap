// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recursiveModuleBundle osmo 1`] = `
"export namespace osmosis {
  export namespace claim {
    export const v1beta1 = {
      ..._177,
      ..._178,
      ..._179,
      ..._180
    };
  }
  export namespace epochs {
    export const v1beta1 = {
      ..._181,
      ..._182
    };
  }
  export namespace gamm {
    export const v1beta1 = {
      ..._183,
      ..._184,
      ..._185,
      ..._186,
      ..._187
    };
  }
  export const incentives = {
    ..._188,
    ..._189,
    ..._190,
    ..._191,
    ..._192
  };
  export const lockup = {
    ..._193,
    ..._194,
    ..._195,
    ..._196
  };
  export namespace mint {
    export const v1beta1 = {
      ..._197,
      ..._198,
      ..._199
    };
  }
  export namespace poolincentives {
    export const v1beta1 = {
      ..._200,
      ..._201,
      ..._202,
      ..._203
    };
  }
  export namespace store {
    export const v1beta1 = {
      ..._204
    };
  }
  export const superfluid = {
    ..._205,
    ..._206,
    ..._207,
    ..._208,
    ..._209
  };
  export namespace txfees {
    export const v1beta1 = {
      ..._210,
      ..._211,
      ..._212,
      ..._213
    };
  }
}"
`;

exports[`recursiveModuleBundle root 1`] = `
"export const osmosis = {
  ..._0
};
export const tendermint = {
  ..._0
};"
`;

exports[`recursiveModuleBundle single 1`] = `
"export const ics23 = {
  ..._0
};"
`;

exports[`recursiveModuleBundle tendermint 1`] = `
"export namespace tendermint {
  export const abci = {
    ..._223
  };
  export const crypto = {
    ..._224,
    ..._225
  };
  export namespace libs {
    export const bits = {
      ..._226
    };
  }
  export const p2p = {
    ..._227
  };
  export const types = {
    ..._228,
    ..._229,
    ..._230,
    ..._231,
    ..._232
  };
  export const version = {
    ..._233
  };
}"
`;
